subxt/src/backend/unstable/follow_stream_unpin.rs

```diff
@@ -109,7 +109,11 @@ impl<Hash: BlockHash> Stream for FollowStreamUnpin<Hash> {
                 FollowStreamMsg::Event(FollowEvent::Initialized(details)) => {
                     // The first finalized block gets the starting block_num.
                     let rel_block_num = this.rel_block_num;
-                    let block_ref = this.pin_block_at(rel_block_num, details.finalized_block_hash);
+                    let block_ref = this.pin_block_at(
+                        rel_block_num,
+                        details.finalized_block_hash,
+                        UnpinPolicy::MaxAge,
```

Comment:
I don't really understand the need for this `UnpinPolicy`; it looks like we are just pinning blocks for as long as possible and no longer unpinning blocks when dropped now?

I was sortof assuming that the change would be smaller; on line https://github.com/paritytech/subxt/pull/1367/files#diff-4ef5f554da480b7c19eac236d6fa5cc67ab6d6d4441eec62765dd55e90df428fR197 we unpin only the blocks listed as pruned blocks rather than unpin _all_ blocks in the unpin set? Perhaps I'm missing an important detail :)

Reply:
To unpin only blocks reported as [pruned](https://github.com/paritytech/subxt/pull/1367/files#diff-4ef5f554da480b7c19eac236d6fa5cc67ab6d6d4441eec62765dd55e90df428fR183) by the finalized event we would need:
- to pass the pruned items of the finalized to the `unpin_blocks`
- and have the pruned blocks be present in the [UnpinFlags](https://github.com/paritytech/subxt/pull/1367/files#diff-4ef5f554da480b7c19eac236d6fa5cc67ab6d6d4441eec62765dd55e90df428fR301) -- their reference is dropped

However, we would never see a `pruned` block present in the `UnpinFlags` because we are pinning them again when we see the `Finalized` event [here](https://github.com/paritytech/subxt/pull/1367/files#diff-4ef5f554da480b7c19eac236d6fa5cc67ab6d6d4441eec62765dd55e90df428fR189). And that would remove their entry from the [UnpinFlags here](https://github.com/paritytech/subxt/pull/1367/files#diff-4ef5f554da480b7c19eac236d6fa5cc67ab6d6d4441eec62765dd55e90df428fR301).

We could move `this.unpin_blocks(cx.waker())` before we pin the prunned blocks. And in doing so cleaning the pruned blocks right away, iff the user has dropped the `NewBlock` reference. However, in doing so, we would return a dummy `BlockRef` for the pruned block that has just been unpinned.

Reply:
Ah interesting, and another thought is that if you haven't yet dropped a block that appears as a pruned block at some point, and then drop it after, then it will never show up in the pruned blocks list again. (In that case though, one could make a note of all currently pinned blocks that have been pruned or something).

I'm not overly comfortable with having most blocks exist just for the max age (in that case I'd lean towards only ever pruning blocks that reach a max age and never trying to do any smarter unpinning). Lemme have a bit of a play too to get a better feel for this and experiment with some other unpinning approach :)

Reply:
We could probably do something similar to the implementation of the `chainHead` in substrate; where we keep track of blocks a bit longer.
That would mean transforming the HashSet for UnpinFlags to a stateful HashMap; where we keep track of unpins from NewBlock refs and Finalized/Pruned refs; and only submit unpin requests when the latter happens. 

Indeed having the blocks live a `max_life` would probably mean that we might get terminated since the substrate implementation is free to change the server `max_life`.

Reply:
Building off your branch here, this is the approach I am trying out:

https://github.com/paritytech/subxt/pull/1368